crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -152,57 +161,308 @@ impl RawModuleResolutionSettings {
+        // gives no such guarantees. Therefore, collecting and then sorting is necessary.
+        // https://docs.python.org/3/library/site.html#module-site
+        let mut all_pth_files: Vec<PthFile> = pth_file_iterator.collect();
+        all_pth_files.sort_by(|a, b| a.path.cmp(&b.path));
+
+        for pth_file in &all_pth_files {
+            dynamic_paths.extend(pth_file.iter_editable_installations().filter_map(
+                |editable_path| {
+                    let possible_search_path = Arc::new(editable_path);
+                    (!static_paths.contains(&possible_search_path)).then_some(possible_search_path)
+                },
+            ));
+        }
+    }
+    dynamic_paths
+}
+
+/// Iterate over the available module-resolution search paths,
+/// following the invariants maintained by [`sys.path` at runtime]:
+/// "No item is added to `sys.path` more than once."
+/// Dynamic search paths (required for editable installs into `site-packages`)
+/// are only calculated lazily.
 ///
-/// [typing spec]: https://typing.readthedocs.io/en/latest/spec/distributing.html#import-resolution-ordering
-#[derive(Clone, Debug, Default, Eq, PartialEq)]
-pub(crate) struct OrderedSearchPaths(Vec<Arc<ModuleResolutionPathBuf>>);
+/// [`sys.path` at runtime]: https://docs.python.org/3/library/site.html#module-site
+struct SearchPathIterator<'db> {
+    db: &'db dyn Db,
+    static_paths: ordermap::set::IntoIter<&'db SearchPathRoot>,
+    dynamic_paths: Option<ordermap::set::Iter<'db, SearchPathRoot>>,
+}
+
+impl<'db> SearchPathIterator<'db> {
+    fn new(db: &'db dyn Db) -> Self {
+        let static_paths = module_resolver_settings(db)
+            .search_path_settings
+            .static_search_paths()
+            .into_iter();
+
+        Self {
+            db,
+            static_paths,
+            dynamic_paths: None,
+        }
+    }
+}
+
+impl<'db> Iterator for SearchPathIterator<'db> {
+    type Item = &'db SearchPathRoot;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let SearchPathIterator {
+            db,
+            static_paths,
+            dynamic_paths,
+        } = self;
+
+        static_paths.next().or_else(|| {
+            dynamic_paths
+                .get_or_insert_with(|| dynamic_module_resolution_paths(*db).into_iter())
+                .next()
+        })
+    }
+}
 
-impl Deref for OrderedSearchPaths {
-    type Target = [Arc<ModuleResolutionPathBuf>];
+impl<'db> FusedIterator for SearchPathIterator<'db> {}
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+fn iter_module_search_paths(db: &dyn Db) -> SearchPathIterator {
+    SearchPathIterator::new(db)
+}
+
+/// Represents a single `.pth` file in a `site-packages` directory.
+/// One or more lines in a `.pth` file may be a (relative or absolute)
+/// path that represents an editable installation of a package.
+struct PthFile<'db> {
+    db: &'db dyn Db,
+    path: SystemPathBuf,
```

Comment:
Can we store just the `File` instead? 

On the other hand, if we use `db.report_untracked_read`, then there's little benefit for storing `File` or even using `SourceText`. You can then simply store the `SystemPathBuf` and directly call `system.read_to_string`.

Reply:
> Can we store just the `File` instead?

Hmm, if I make this change (which I think is what you're suggesting):

```diff
--- a/crates/red_knot_module_resolver/src/resolver.rs
+++ b/crates/red_knot_module_resolver/src/resolver.rs
@@ -330,7 +330,7 @@ impl<'db> FusedIterator for SearchPathIterator<'db> {}
 struct PthFile<'db> {
     db: &'db dyn Db,
     path: SystemPathBuf,
-    contents: SourceText,
+    file: File,
     site_packages: &'db SystemPath,
 }
 
@@ -344,7 +344,7 @@ impl<'db> PthFile<'db> {
         Self {
             db,
             path: pth_path,
-            contents: source_text(db.upcast(), file),
+            file,
             site_packages,
         }
     }
@@ -355,11 +355,13 @@ impl<'db> PthFile<'db> {
     where
         'a: 'db,
     {
+        let contents = source_text(self.db.upcast(), self.file);
+
         // Empty lines or lines starting with '#' are ignored by the Python interpreter.
         // Lines that start with "import " or "import\t" do not represent editable installs at all;
         // instead, these are files that are executed by Python at startup.
         // https://docs.python.org/3/library/site.html#module-site
-        self.contents.lines().filter_map(|line| {
+        contents.lines().filter_map(|line| {
             let line = line.trim();
             if line.is_empty()
                 || line.starts_with('#')
```

Then I end up with compile errors:

```
error[E0515]: cannot return value referencing local variable `contents`
   --> crates/red_knot_module_resolver/src/resolver.rs:364:9
    |
364 |           contents.lines().filter_map(|line| {
    |           ^-------
    |           |
    |  _________`contents` is borrowed here
    | |
365 | |             let line = line.trim();
366 | |             if line.is_empty()
367 | |                 || line.starts_with('#')
...   |
374 | |             ModuleResolutionPathBuf::editable_installation_root(self.db, possible_editable_install)
375 | |         })
    | |__________^ returns a value referencing data owned by the current function
    |
    = help: use `.collect()` to allocate the iterator
```

Which I could fix easily enough by creating a custom iterator type that stores value returned by `source_text()`. But that doesn't seem much different to storing the value returned by `source_text()` directly on `PthFile` itself, which is what I have now.

> On the other hand, if we use `db.report_untracked_read`, then there's little benefit for storing `File` or even using `SourceText`. You can then simply store the `SystemPathBuf` and directly call `system.read_to_string`.

I tried this, but it makes `deleting_pth_file_on_which_module_resolution_depends_invalidates_cache()` start failing again (it was fixed by adding the `report_untracked_read()` call).

Reply:
I think I got a little bit closer to what you were recommending in https://github.com/astral-sh/ruff/pull/12307/commits/0c1e24c07dd0b742f8796243588f335be225a701

Reply:
I switched it to use `String`.